### PR TITLE
Padding

### PR DESCRIPTION
**Pull request by nilgoyette**
This PR modifies the `pad` function to

- accept n-dimensional arrays (instead of only 3D)
- Add almost all NumPy pad modes. Only `linear_ramp` and `<function>` are missing. I do not expect to add them now or ever.
- Add several tests for all modes

I believe it is much cleaner than NumPy version but it is also slower :( In fact, we're around 2-4x slower than NumPy, depending on the mode. See #4 for more information.

The `ndarray-stats` dependency was added.

### CODE REVIEW

File: src/pad.rs [hunk 1]
```diff
@@ -1,112 +1,213 @@
...
-    let size = size as u16;
-    let padding = padding as u16;
-    v.extend(size - padding..size);
-    v.extend(0..size);
-    v.extend(0..padding);
-    v
-}
+    // Select portion of padded array that needs to be copied from the original array.
+    let mut orig_portion = padded.view_mut();
+    for d in 0..data.ndim() {
```
Comment (czotti) [inline src/pad.rs:162]: Could you use https://docs.rs/ndarray/latest/ndarray/struct.ArrayBase.html#method.slice_each_axis_mut
for this.

**Review by czotti — COMMENTED** [review]
Also as a general comment I think we could use https://docs.rs/ndarray/latest/ndarray/struct.ArrayBase.html#method.multi_slice_mut
To try to simplify the data manipulation no ?

I have a feeling that we could use better slicing but I can't tell how to make it work.

To the best of my knowledge this looks fine. GJ Nil.

File: src/pad.rs (hunk 1 above)
Comment (nilgoyette) [reply src/pad.rs:162]: Haha, wow, good find! Yes, `slice_each_axis_mut` replaces the manual loop.

**Comment by nilgoyette** [comment]
> Also as a general comment I think we could use https://docs.rs/ndarray/latest/ndarray/struct.ArrayBase.html#method.multi_slice_mut To try to simplify the data manipulation no ?

Using `multi_slice_mut` might be able to replace my 2 buffers.
- One is used with `Median`. I can't avoid it because I must sort the data.
- The other is used in `ByIndices`. It looks like it could be replaced but I'm sure it can't. We're only writing in the left and right part, but we're reading the whole lane. `multi_slice_mut` must be able to prove that both slices don't touch. They do touch.

**Pull request merged**